osu.Game/Rulesets/Edit/Checks/CheckDelayedHitsounds.cs

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using osu.Framework.Audio.Track;
+using osu.Game.Audio;
+using osu.Game.Extensions;
+using osu.Game.Rulesets.Edit.Checks.Components;
+
+namespace osu.Game.Rulesets.Edit.Checks
+{
+    public class CheckDelayedHitsounds : ICheck
+    {
+        /// <summary>
+        /// Threshold at which point the sample is considered silent.
+        /// </summary>
+        private const float silence_threshold = 0.001f;
+
+        private const float falloff_factor = 0.95f;
+        private const int delay_threshold = 5;
+        private const int delay_threshold_negligible = 1;
+
+        public CheckMetadata Metadata => new CheckMetadata(CheckCategory.Audio, "Delayed hit sounds.");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateConsequentDelay(this),
+            new IssueTemplateDelay(this),
+            new IssueTemplateDelayNoSilence(this),
+            new IssueTemplateMinorDelay(this),
+            new IssueTemplateMinorDelayNoSilence(this),
+        };
+
+        private float getAverageAmplitude(Waveform.Point point) => (point.AmplitudeLeft + point.AmplitudeRight) / 2;
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            var beatmapSet = context.Beatmap.BeatmapInfo.BeatmapSet;
+
+            if (beatmapSet == null)
+                yield break;
+
+            foreach (var file in beatmapSet.Files)
+            {
+                using (Stream? stream = context.WorkingBeatmap.GetStream(file.File.GetStoragePath()))
+                {
+                    if (stream == null)
+                        continue;
+
+                    if (!isHitSound(file.Filename))
+                        continue;
+
+                    using (Waveform waveform = new Waveform(stream))
```

Comment:
I'd actually recommend to switch to `using var waveform = new Waveform(stream);` here to reduce the nesting.

The outer stream using I'd probably leave as an explicit using block because with streams the using declaration thing gets a bit foot-gun-y, but with this there should be no harm.